File: src/main/java/com/tomergabel/examples/java/PatternMatching1.java
package com.tomergabel.examples.java;

import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.Scanner;

public class PatternMatching1 {

    static final int DEFAULT_LINE_COUNT = 10;

    public static void main(String[] args) throws IOException {
        // CLI --

        InputStream in = System.in;
        int lines = DEFAULT_LINE_COUNT;

        if (args.length > 0)
            in = new FileInputStream(args[0]);
        if (args.length > 1)
            lines = Integer.valueOf(args[1]);
        if (args.length > 2) {
            System.out.println("Usage: sbt 'runMain " + PatternMatching1.class.getName() + " [input] [lines]'");
            System.exit(-1);
        }

        // Test --

        Scanner scanner = new Scanner(in);
        try {
            int line = 0;
            while (scanner.hasNextLine() && line < lines) {
                System.out.println(scanner.nextLine());
                line++;
            }
        } finally {
            in.close();
        }
    }
}

File: src/main/scala/com/tomergabel/examples/patterns/PatternMatching1.scala
package com.tomergabel.examples.patterns

import java.io.FileInputStream
import java.util.Scanner

/**
  * Created by tomerga on 04/04/2016.
  */
object PatternMatching1 extends App {

  val defaultLineCount = 10

  // CLI --

  val (in, lines) =
    args match {
      case Array() =>
        (System.in, defaultLineCount)

      case Array(inPath) =>
        (new FileInputStream(inPath), defaultLineCount)

      case Array(inPath, lineStr) =>
        (new FileInputStream(inPath), lineStr.toInt)

      case _ =>
        println(s"Usage: sbt 'runMain ${this.getClass.getName} [input] [lines]'")
        sys.exit(-1)
    }

  // Test --

  val scanner = new Scanner(in)
  try {
    var line = 0
    while (scanner.hasNextLine && line < lines) {
      println(scanner.nextLine)
      line = line + 1
    }
  } finally {
    in.close()
  }
}
